src/Compilers/CSharp/Test/Emit3/Diagnostics/DiagnosticAnalyzerTests.cs

```diff
@@ -4441,5 +4442,203 @@ record B(int I) : A(I);
+                context.RegisterSyntaxTreeAction(context => { AssertSame(context.Options); RegisterSyntaxTreeActionInvoked = true; });
+
+                context.RegisterOperationBlockStartAction(context =>
+                {
+                    AssertSame(context.Options);
+                    RegisterOperationBlockStartActionInvoked = true;
+                    context.RegisterOperationBlockEndAction(context =>
+                    {
+                        AssertSame(context.Options);
+                        RegisterOperationBlockEndActionInvoked = true;
+                    });
+                });
+
+                context.RegisterCompilationStartAction(context =>
+                {
+                    AssertSame(context.Options);
+                    RegisterCompilationStartActionInvoked = true;
+                    context.RegisterCompilationEndAction(context =>
+                    {
+                        AssertSame(context.Options);
+                        RegisterCompilationEndActionInvoked = true;
+                    });
+                });
+                context.RegisterSymbolStartAction(context =>
+                {
+                    AssertSame(context.Options);
+                    RegisterSymbolStartActionInvoked = true;
+                    context.RegisterSymbolEndAction(context =>
+                    {
+                        AssertSame(context.Options);
+                        RegisterSymbolEndActionInvoked = true;
+                    });
+                }, SymbolKind.NamedType);
+            }
+
+            public void AssertAllCallbacksInvoked()
+            {
+                Assert.NotNull(SeenOptions);
+
+                Assert.True(RegisterAdditionalFileActionInvoked);
+
+                Assert.True(RegisterAdditionalFileActionInvoked);
+                Assert.True(RegisterCodeBlockActionInvoked);
+                Assert.True(RegisterCodeBlockStartActionInvoked);
+                Assert.True(RegisterCompilationActionInvoked);
+                Assert.True(RegisterOperationActionInvoked);
+                Assert.True(RegisterOperationBlockActionInvoked);
+                Assert.True(RegisterSemanticModelActionInvoked);
+                Assert.True(RegisterSymbolActionInvoked);
+                Assert.True(RegisterSyntaxNodeActionInvoked);
+                Assert.True(RegisterSyntaxTreeActionInvoked);
+
+                Assert.True(RegisterOperationBlockStartActionInvoked);
+                Assert.True(RegisterOperationBlockEndActionInvoked);
+                Assert.True(RegisterCompilationStartActionInvoked);
+                Assert.True(RegisterCompilationEndActionInvoked);
+                Assert.True(RegisterSymbolStartActionInvoked);
+                Assert.True(RegisterSymbolEndActionInvoked);
+            }
+        }
+
+        [Fact]
+        public async Task TestAnalyzerSpecificOptionsFactory()
```

Comment:
This test ensures that *all* callbacks get the right options instance if customized by teh host.